### PR TITLE
Add Point function, so that string feature can use it instead of line…

### PR DESCRIPTION
**Pull request by jq**
…ar function

@deerzq @saurfang

### CODE REVIEW

File: core/src/main/java/com/airbnb/aerosolve/core/function/Linear.java [hunk 1]
```diff
@@ -98,6 +98,7 @@ public void LInfinityCap(float cap) {
 
   @Override
   public float LInfinityNorm() {
+    // return max contribution
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Linear.java]: max absolute contribution

File: core/src/main/java/com/airbnb/aerosolve/core/function/MultiDimensionSpline.java [hunk 2]
```diff
@@ -247,10 +247,9 @@ public void smooth(double tolerance) {
     }
   }
 
-  // TODO: this is a shallow copy. implement deep copy if need immutability
   @Override
   public MultiDimensionSpline clone() throws CloneNotSupportedException {
-    return (MultiDimensionSpline) super.clone();
+    return new MultiDimensionSpline(this.ndTreeModel);
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/MultiDimensionSpline.java]: why this change?

File: core/src/main/java/com/airbnb/aerosolve/core/function/Point.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    this.weight = weight;
+  }
+
+  public Point(ModelRecord record) {
+    List<Double> weightVec = record.getWeightVector();
+    weight = weightVec.get(0).floatValue();
+  }
+
+  @Override
+  public Function aggregate(Iterable<Function> functions, float scale, int numBins) {
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Point.java]: not for this PR, but `numBins` is not used, I feel we may want to make `numBins` available only to Spline related features, otherwise, whenever we add new function type, we have to add this.

File: core/src/main/java/com/airbnb/aerosolve/core/function/Point.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    float aggWeight = 0;
+    for (Function fun: functions) {
+      Point point = (Point) fun;
+      aggWeight += scale * point.weight;
+    }
+    return new Point(aggWeight);
+  }
+
+  @Override
+  public float evaluate(float... x) {
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Point.java]: for string feature, we assume `x=1`, probably we should return `weight * x` so that using `Point` to represent features in AdditiveModel would be make this additive model equivalent to conventional linear model.

Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Point.java]: By doing this, `Point` can also be used for other float features, not just string features

File: core/src/main/java/com/airbnb/aerosolve/core/function/MultiDimensionSpline.java (hunk 2 above)
Comment (jq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/MultiDimensionSpline.java]: this is the correct way to clone it

File: core/src/main/java/com/airbnb/aerosolve/core/function/Point.java (hunk 3 above)
Comment (jq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Point.java]: this is to hide the implementation details

File: core/src/main/java/com/airbnb/aerosolve/core/function/Point.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+  @Override
+  public void LInfinityCap(float cap) {
+    if (cap <= 0.0f) return;
+    weight = Math.min(cap, weight);
+  }
+
+  @Override
+  public float LInfinityNorm() {
+    return weight;
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/function/Point.java]: `Math.abs(weight)`, the definition of `LInfinityNorm` won't be consistent with the definition in `Linear` and `Spline` unless we also add `minVal` and `maxVal` as member variables, then the norm would be `Math.max(Math.abs(weight * minVal), Math.abs(weight * maxVal))`. But for the current string feature use case, `Math.abs(weight)` would be good. Perhaps add a `TODO` here.

**Comment by deerzq** [comment]
LGTM
